### PR TITLE
Refactor cast storage

### PR DESCRIPTION
**Pull request by reminisce**
Moves cast_storage operator related functions out of elemwise_unary_op.h, .cc, and .cu to cast_storage-inl.h, cast_storage.cc and cast_storage.cu.

### CODE REVIEW

File: src/operator/nn/cast_storage.cc [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+#include "../elemwise_op_common.h"
+#include "../tensor/elemwise_unary_op.h"
+
+namespace mxnet {
+namespace op {
+
+// TODO(haibin) declare backward op for cast storage
+// Only support cast to default storage now
+// Other types require add infer_storage type pass
+DMLC_REGISTER_PARAMETER(CastStorageParam);
```
Comment (eric-haibin-lin) [inline src/operator/nn/cast_storage.cc]: Remove 
```
+// Only support cast to default storage now
+// Other types require add infer_storage type pass
``

File: src/operator/nn/cast_storage.cc [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+.set_num_inputs(1)
+.set_num_outputs(1)
+.set_attr_parser(ParamParser<CastStorageParam>)
+.set_attr<nnvm::FInferShape>("FInferShape", ElemwiseShape<1, 1>)
+.set_attr<nnvm::FInferType>("FInferType", ElemwiseType<1, 1>)
+.set_attr<nnvm::FInferStorageType>("FInferStorageType", CastStorageInferStorageType)
+.set_attr<FCompute>("FCompute<cpu>", IdentityCompute<cpu>)
+// _backward pass
+// .set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseNone{"negative"})
+.set_attr<FComputeEx>("FComputeEx<cpu>", CastStorageComputeEx<cpu>)
```
Comment (eric-haibin-lin) [inline src/operator/nn/cast_storage.cc]: remove 
```
+// _backward pass
 +// .set_attr<nnvm::FGradient>("FGradient", ElemwiseGradUseNone{"negative"})
```

File: src/common/utils.h [hunk 3]
```diff
@@ -152,6 +153,16 @@ inline int GetExecNumMatchColor() {
...
+template<typename T, typename V>
+V ParallelAccumulate(const T* a, const size_t n, V start) {
+  V sum = start;
+#pragma omp parallel for reduction(+:sum)
+  for (size_t i = 0; i < n; ++i) {
+    sum += a[i];
+  }
+  return sum;
+}
+
```
Comment (eric-haibin-lin) [inline src/common/utils.h]: why are there two types T and V?

File: src/operator/nn/cast_storage-inl.h [hunk 4]
```diff
@@ -57,23 +57,41 @@ inline void CastStorageDnsRspImpl(mshadow::Stream<cpu>* s, const TBlob& dns, NDA
...
       mxnet_op::Kernel<MarkRspRowIdx, cpu>::Launch(s, num_rows, row_idx,
           dns.dptr<DType>(), num_cols);
+      double elapsed1 = dmlc::GetTime() - start;
+
       index_t nnr = 0;
-      nnr = std::accumulate(row_idx, row_idx+num_rows, nnr);
+      start = dmlc::GetTime();
+      nnr = mxnet::common::ParallelAccumulate(row_idx, num_rows, nnr);
+      //nnr = std::accumulate(row_idx, row_idx+num_rows, nnr);
+      double elapsed2 = dmlc::GetTime() - start;
```
Comment (eric-haibin-lin) [inline src/operator/nn/cast_storage-inl.h]: remove it ? 
`+      //nnr = std::accumulate(row_idx, row_idx+num_rows, nnr);`

File: src/operator/nn/cast_storage-inl.h [hunk 5]
```diff
@@ -57,23 +57,41 @@ inline void CastStorageDnsRspImpl(mshadow::Stream<cpu>* s, const TBlob& dns, NDA
...
       }
+      double elapsed3 = dmlc::GetTime() - start;
+
+      double total = elapsed1 + elapsed2 + elapsed3;
+      LOG(INFO) << "shape = " << rsp->shape();
+      LOG(INFO) << "nnr = " << nnr;
+      LOG(INFO) << "MarkRspRowIdx cost " << elapsed1 * 1000 << " ms, " << static_cast<int>(elapsed1/total*100) << '%';
+      LOG(INFO) << "ParallelAccumulate cost " << elapsed2 * 1000 << " ms, " << static_cast<int>(elapsed2/total*100) << '%';
+      LOG(INFO) << "Copy rows cost " << elapsed3 * 1000 << " ms, " << static_cast<int>(elapsed3/total*100) << '%';
     });
```
Comment (eric-haibin-lin) [inline src/operator/nn/cast_storage-inl.h]: We're not committing these profiling log, are we?

File: src/operator/nn/cast_storage-inl.h (hunk 4 above)
Comment (reminisce) [inline src/operator/nn/cast_storage-inl.h]: Yes, I will push another commit to clean up everything. This is kept here for convenience for now.

File: src/operator/nn/cast_storage-inl.h (hunk 5 above)
Comment (reminisce) [inline src/operator/nn/cast_storage-inl.h]: No, we are not. I will push another commit to clean up everything later.

File: src/common/utils.h (hunk 3 above)
Comment (reminisce) [inline src/common/utils.h]: T is for the element type of the array, and V is for the type of the accumulated sum.

**Pull request merged**